common/types/encode.go

```diff
@@ -10,7 +10,7 @@ import (
 )
 
 // Bytes returns the BlockID as a byte slice.
-func (id BlockID) Bytes() []byte { return id.AsHash32().Bytes() }
+func (id BlockID) Bytes() []byte { return id.AsHash20().Bytes() }
```

Comment:
this is not correct I think, and will interfere with current sync and DBconventions. 
currently all IDs in all DBs are of Hash32 format and i think we should keep it as such since it unified our ID in DB format

Reply:
I'm not sure you're right about this.

Database keys are variable-size byte slices and the blocks database only has blocks in it, so there's no advantage for conformity there. The only way to fetch from this database is by using a getter that accepts a `BlockID` as an argument (not a byte slice).

The sync protocol requires a `Hash32` as a key and it still gets it (justifiably doesn't use `.Bytes()`). The block request handler converts the `Hash32` it receives to a `Hash20` and then to a `BlockID` before fetching from the database anyway (this was before my change).

All the tests are passing.

Reply:
BTW, if we really do need to often get the padded 32-byte representation (I don't think we do) then we should create a method named `.Bytes32()` which would be explicit and less confusing than padding the returned byte slice from `.Bytes()`.

Reply:
the new sync relies on the keys of all DBs to be the size of hash32

Reply:
Can you be more specific? It's hard for me to respond on such a high level...

Did you make the new sync access the blocks database directly, not via the setter and getter as it is in `develop` now?

Can you send me a link to the code so I can take a look and understand?

Reply:
This is the current sync block request handler:

https://github.com/spacemeshos/go-spacemesh/blob/f76c8b99427a4a9f559da87fced6cf5aa3219266/sync/handler.go#L89-L123

As you can see, the list of `blockids` it receives is a list of `Hash32`s, which it then converts to `BlockID`s before fetching the blocks.

What's the equivalent code that is sensitive to the format of the database key?

Reply:
this is the new fetch code, as you can see it is generic and assumes same size of hash for all hashes
https://github.com/spacemeshos/go-spacemesh/pull/2213/commits/d1aa40236595002e4541db8b46c315c7ef937dd5#diff-f03683d1546e71521624eea58e910c92a2525ec1a0c6d216c8cad6b3a9d26a8aR252-R299

Reply:
Yes, this code will break that design.

It seems we're already too far along on that path to discuss if it's a good idea or not. I intuitively don't like allowing remote queries directly into the node's databases. I need to give it some thought if it's a security issue or not (given that all the data in these DBs is public anyway). That's regardless this specific fix.

In order to simplify the discussion I'll create an alternative fix for Aviv's issue that's local to the API rather than fixing (what I consider to be) the underlying problem.

Reply:
Assuming we stay with your current design, I think we should introduce a `.Bytes32()` method and implement that on all key objects. This will make it explicit that this method always returns 32 bytes, and not necessarily the actual bytes of the object - I think it's currently confusing.

I also suggest creating a wrapper of our current database wrapper, where the `.Get()` and `.Put()` methods take a `Hash32` as the key - I'd then make the sync code only work with that kind of database and not the generic version where keys can be any length. I think the current implementation is possibly confusing and introduces unnecessary risk of coding errors.